Remove deprecated "_method" requirement.

tackerm opened the pull request:
I am targeting this branch, because I simply don't know if it is a BC break or not.

Closes #437

## Changelog

<!-- MANDATORY
    Fill the changelog part inside the code block.
    Follow this schema: http://keepachangelog.com/
-->

<!-- REMOVE EMPTY SECTIONS -->
```markdown
### Removed
- Removed deprecated `_method` requirement from RouteCollection.
```


## Subject
Remove deprecated "_method" requirement.

greg0ire commented:
The right way to do this is to detect the current version of whatever sf component is complaining, and remove the method argument based on that. This would be BC. You should 

- `git reset --hard origin/1.x`
- `git cherry-pick b772fcfdc69d37f09ef491e9448a1ef84193bbb1`
- `git push --force`

I changed the base branch to 1.x for you.

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

tackerm commented:
I should really stop patching stuff in the web editor... :man_facepalming: 

@greg0ire Thanks for the hints. I did as requested, fixed the typo, squashed commits... the PR seems not to get updated though. Instead of messing around with rebasing and merge conflicts, maybe I should just create a new, clean PR against 1.x for simplicity?

greg0ire commented:
> maybe I should just create a new, clean PR against 1.x for simplicity?

Please don't, please try the commands I provided you with, and show me the output.

tackerm commented:
I did that, the changes are in the 1.x branch of my fork. But that doesn't update the PR as it is based on my master branch...

greg0ire commented:
Okay, then maybe "origin" is your fork here? Here are new instructions that are less error prone:

1. `git reset --hard 877431e9a2958ab6aecc3ff3492979cd0cd5fc7f`
2. `git cherry-pick b772fcfdc69d37f09ef491e9448a1ef84193bbb1`
3. `git push --force`

tackerm commented:
Well yes, I am working on my fork. The problem is that the PR now tries to merge my fork:master -> upstream:1.x - that will obviously not work. Instead it should be a PR fork:1.x -> upstream:1.x

greg0ire commented:
Naming it master was indeed not a great idea. It will be confusing, but will work though. Just please follow the step, and we will just ignore the name, it's really not an issue.

greg0ire commented:
If this still does not work for you, then create a new PR, but this time, please name your branch something else than 1.x or master ;)

tackerm commented:
Done - I didn't _name_ it master, I just forked and made the changes via the web ui - on the actual master branch of my fork :see_no_evil: 

Anyways, I have now hard-resetted my master (never thought I'd do anything like that), and it looks good so far...

greg0ire commented:
Oh right, when you edit a file from the main repo, it would have named it `patch-1` or something like that. When I develop on a project, I just delete the master  branch on my fork :) You should never need it ever :P

greg0ire commented:
Ok so now you probably need to read [this](https://symfony.com/doc/current/setup/bundles.html#updating-your-code-to-support-symfony-2-x-and-3-x-at-the-same-time)

tackerm commented:
Ok, we have already determined that I am a bit slow today, but: The 1.x branch requires `"symfony/framework-bundle": "^2.3"` which in turn requires `"symfony/routing": "~2.3"`. The "_method" requirement was already deprecated in 2.2, so this should be compatible with all currently supported versions...?

greg0ire commented:
> Ok, we have already determined that I am a bit slow today

You're actually doing great, keep up the good work!

> The "_method" requirement was already deprecated in 2.2, so this should be compatible with all currently supported versions...?

Was it? if yes then indeed, no need for detection/conditions. Can you give the source for this in your commit message?

tackerm commented:
The [deprecation notice added in 2.7](https://github.com/symfony/routing/blob/2.7/Route.php#L652) says so and the [actual code in 2.2](https://github.com/symfony/routing/blob/2.2/Route.php#L585) confirms that.

The details are in #437

greg0ire commented on Route/PathInfoBuilderSlashes.php:
```diff
@@ -26,11 +26,11 @@ public function build(AdminInterface $admin, RouteCollection $collection)
         $collection->add('list');
         $collection->add('create');
-        $collection->add('batch', null, array(), array('_method' => 'POST'));
+        $collection->add('batch', null, array());
```
There is [a dedicated argument for methods](https://github.com/sonata-project/SonataAdminBundle/blob/3.x/Route/RouteCollection.php#L70), please use it.

tackerm replied:
THAT was the part I missed :)

greg0ire commented:
I removed your TODO list

greg0ire commented:
@sonata-project/contributors please review

dbu commented on Route/PathInfoBuilderSlashes.php:
```diff
@@ -26,11 +26,12 @@ public function build(AdminInterface $admin, RouteCollection $collection)
         $collection->add('list');
         $collection->add('create');
-        $collection->add('batch', null, array(), array('_method' => 'POST'));
+        $collection->add('batch', null, array(), array(), array(), '', array(), array('POST'));
```
i think this only works by accident because php allows extra parameters and because by default a route allows all methods. the signature of [RouteCollection::add in version 2](https://github.com/sonata-project/SonataAdminBundle/blob/2.x/Route/RouteCollection.php#L66) does not have the extra arguments like version 3 has. i think for correct BC we need a version check and duplicate the above line, until we drop support for sonata admin 2.*

greg0ire replied:
> until we drop support for sonata admin 2.*

I forgot you still supported it, sorry. You should consider dropping it, it must hinder development a lot.

dbu replied:
the upcoming 2.x release (master branch) does drop it and goes to ^3.1

tackerm replied:
This looks a bit tricky: SonataAdminBundle requires [symfony/routing : ~2.3](https://github.com/sonata-project/SonataAdminBundle/blob/2.x/composer.json#L25) - the "_method" requirement was already deprecated in 2.2, but sonata admin does not seem to provide any alternative to set the route methods.

Implementing a BC check here would therefore not fix the deprecation for sonata admin 2.x - which means we might as well merge this into master?

greg0ire replied:
sonata admin 2 is not compatible with sf 3 AFAIK, which means you can leave the code as it was before for Sonata 2.x

dbu replied:
@tackerm indeed you are right here. i will cherry-pick your commit to the master branch.

greg0ire replied:
Oh right I overlook the `master` part! Instead of cherry-picking, maybe just change the base?

greg0ire replied:
Changing the base shows that master is not up to date with 3.x. You may get rid of the offending commits by using interactive rebase. Or you can just wait for the cherry-pick.

dbu commented:
committed to master in d1ca0fb5b35b08cc9fa1ba5a2d7e2b551b0db617

thanks a lot @tackerm !
